### PR TITLE
Scheduler Rework

### LINKED ISSUE #576: Turn schedulers into function-like objects
Scheduler's are currently normal functions. Turning them into function-like objects has the main benefit of allowing them to be serialized with the model using the `AgentsIO` submodule. It would end up being something like this:

```julia
abstract type AbstractScheduler end

struct Fastest <: AbstractScheduler end

(::Fastest)(model::ABM) = keys(model.agents)
```

and so on. It also makes schedulers that store data a bit easier, so `partially` can be

```julia
struct Partially <: AbstractScheduler
    p::Float64
end

(sched::Partially)(model::ABM) = randomsubseq(model.rng, collect(keys(model.agents)), sched.p)
```
It's a breaking change, so should be good to get in before v5

### PR DESCRIPTION
**Pull request by AayushSabharwal**
Closes #576

Changing `ByType` to returning an iterator instead of `vcat(sets...)` should help with allocations (#581)

TODO:
- [ ] `UserScheduler`
- [ ] Global search and replace for names of schedulers
- [ ] Fix JLD2
  - [ ] Warnings for `UserScheduler`
- [ ] Docs
- [ ] Tests
- [ ] Examples

### CODE REVIEW

**Comment by AayushSabharwal** [comment]
I'm not entirely sure if `ByType` will serialize properly, but I think it should.

File: src/submodules/schedulers.jl [hunk 1]
```diff
@@ -28,117 +28,108 @@ module Schedulers
...
 using Random: shuffle!, randsubseq
 
-export randomly, by_id, fastest, partially, by_property, by_type
+export AbstractScheduler,
+    Fastest,
+    ByID,
+    Randomly,
+    Partially,
+    ByProperty,
+    ByType
```
Comment (Datseris) [inline src/submodules/schedulers.jl:37]: All the old names need to be put in the deprecations! Also warnings need to be thrown in the ByType and ByProperty when used with a function!

**Comment by Datseris** [comment]
@AayushSabharwal I'm really thinking whether such a large change is worth it... User will always have to provide some functions after load, and will again be the same for any advanced scheduler such as user defined. The load function anyways has a default scheduler value that is the same as the default scheduler of `ABM`. So, maybe we are changing too much for too little gain.

**Comment by AayushSabharwal** [comment]
After beginning to write this, I think I see what you mean now. It ends up with too many edge cases. `ByProperty` might end up with a function, which means that needs to be handled separately. Unsure if `ByType` will work. If the user has a custom scheduler that uses a function, they'd need to implement `from_serializable`. It's leading to more code instead of less.

**Comment by Datseris** [comment]
Yeah, let's close this. It was a valid attempt but unless JLD2 gets to know how to save functions, we're not really in a good situation. BSON can save functions but I guess it has many other drawbacks versus JLD2.

**Pull request closed without merging**